cpp/src/arrow/array/data.h

```diff
@@ -434,6 +434,21 @@ struct ARROW_EXPORT ArraySpan {
     return GetValues<T>(i, this->offset);
   }
 
+  // Access a buffer's data as a span
+  template <typename T>
+  util::span<const T> GetSpan(int i) const {
+    return util::span(buffers[i].data_as<T>(),
+                      static_cast<size_t>(buffers[i].size) / sizeof(T))
+        .subspan(offset);
```

Comment:
I'm not sure why, but clang-10 is failing to recognize span as a struct.
https://github.com/apache/arrow/actions/runs/6832493029/job/18583237204?pr=38027#step:7:1342

Could you please rewrite this minimally with a local span and explicit template arguments to see if that satisfies clang-10?

Reply:
Yes I'm curious about that too. I will try it in the next few days.